[Yarr] Improve processing of an alternation of strings

msaboff opened the pull request:
#### 12c34ef5e3058990f7e9997bc07ea1883c2a9eab
<pre>
[Yarr] Improve processing of an alternation of strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=288102">https://bugs.webkit.org/show_bug.cgi?id=288102</a>
<a href="https://rdar.apple.com/145222010">rdar://145222010</a>

Reviewed by Yusuke Suzuki.

Added the notion of a string list to a parsed RegExp that is in the form of
  /^(?:break|case|which|do|for)/ with an optional trailing $.
Such a RegExp will not backtrack and therefore we can streamline the code we emit for such a pattern.

This change involves recognizing beginning of string anchored alternations of strings while parsing and
then treating the generation of JIT code differently for these patterns.  This includes changing how
conditional branching works, specifically that instead of the &quot;fall through on match&quot; for each term,
to a &quot;jump on match&quot; for the whole alternation.

The current code generated for the &quot;case&quot; elternative is:
   8:Term PatternCharacter checked-offset:(3) &apos;c&apos;
               &lt;156&gt; 0x11381430c:    add      w1, w1, #2
               &lt;160&gt; 0x113814310:    cmp      w1, w2
               &lt;164&gt; 0x113814314:    b.hi     0x113814444 -&gt; &lt;468&gt;
  10:Term PatternCharacter checked-offset:(4) &apos;c&apos;
               &lt;168&gt; 0x113814318:    sub      x17, x0, #4
               &lt;172&gt; 0x11381431c:    ldr      w17, [x17, x1]
               &lt;176&gt; 0x113814320:    movz     w16, #0x6163
               &lt;180&gt; 0x113814324:    movk     w16, #0x6573, lsl #16 -&gt; 0x65736163
               &lt;184&gt; 0x113814328:    cmp      w17, w16
               &lt;188&gt; 0x11381432c:    b.ne     0x113814444 -&gt; &lt;468&gt;
  11:Term PatternCharacter checked-offset:(4) &apos;a&apos; already handled
  12:Term PatternCharacter checked-offset:(4) &apos;s&apos; already handled
  13:Term PatternCharacter checked-offset:(4) &apos;e&apos; already handled
  14:NestedAlternativeNext minimum-size:(5),checked-offset:(5)
               &lt;192&gt; 0x113814330:    movz     x16, #0x4444
               &lt;196&gt; 0x113814334:    movk     x16, #0x1381, lsl #16
               &lt;200&gt; 0x113814338:    movk     x16, #0x8001, lsl #32
               &lt;204&gt; 0x11381433c:    movk     x16, #0xc973, lsl #48 -&gt; 0x113814444 JIT PC
               &lt;208&gt; 0x113814340:    stur     x16, [sp, #8]
               &lt;212&gt; 0x113814344:    b        0x113814404 -&gt; &lt;404&gt;
With some additional backtracking code:
   9:NestedAlternativeNext minimum-size:(4),checked-offset:(4)
               &lt;468&gt; 0x113814444:    sub      w1, w1, #2
               &lt;472&gt; 0x113814448:    b        0x113814348 -&gt; &lt;216&gt;

With this change, the processing of &quot;case&quot; becomes:
   9:StringListAlternativeNext minimum-size:(4),checked-offset:(4)
               &lt;132&gt; 0x12a8285c4:    sub      w1, w1, #1
               &lt;136&gt; 0x12a8285c8:    cmp      w1, w2
               &lt;140&gt; 0x12a8285cc:    b.hi     0x12a8285e8 -&gt; &lt;168&gt;
  10:Term PatternCharacter checked-offset:(4) &apos;c&apos;
               &lt;144&gt; 0x12a8285d0:    sub      x17, x0, #4
               &lt;148&gt; 0x12a8285d4:    ldr      w17, [x17, x1]
               &lt;152&gt; 0x12a8285d8:    movz     w16, #0x6163
               &lt;156&gt; 0x12a8285dc:    movk     w16, #0x6573, lsl #16 -&gt; 0x65736163
               &lt;160&gt; 0x12a8285e0:    cmp      w17, w16
               &lt;164&gt; 0x12a8285e4:    b.eq     0x12a82866c -&gt; &lt;300&gt;
  11:Term PatternCharacter checked-offset:(4) &apos;a&apos; already handled
  12:Term PatternCharacter checked-offset:(4) &apos;s&apos; already handled
  13:Term PatternCharacter checked-offset:(4) &apos;e&apos; already handled
  14:StringListAlternativeNext minimum-size:(5),checked-offset:(5)
With no backtracking code.

We are able to eliminate one branch and the saving of the continuation PC for backtracking.
The code size to process these string list RegExp is reduces.  For the example RegExp above,
the prior version created 1940 bytes (485 instructions) of code while the code created with this
1392 bytes (345 instructions) of code, a nearly 30% reduction in code.

This change is a ~18% progression on the new regexp-keyword-parsing microbenchmark:

                                 Baseline               YarrStringList

regexp-keyword-parsing      136.7065+-0.9807     ^    116.0161+-1.1791        ^ definitely 1.1783x faster

&lt;geometric&gt;                 136.7065+-0.9807     ^    116.0161+-1.1791        ^ definitely 1.1783x faster

* JSTests/microbenchmarks/regexp-keyword-parsing.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
(let.re.break.case.catch.continue.debugger.default.else.finally.if):
(let.re1.break.case.catch.continue.debugger.default.else.finally.if):
* JSTests/stress/regexp-parsing-tokens.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::atomParenthesesEnd):
(JSC::Yarr::YarrPatternConstructor::checkForTerminalParentheses):
(JSC::Yarr::PatternAlternative::dump):
(JSC::Yarr::PatternTerm::dump):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::PatternTerm::PatternTerm):
(JSC::Yarr::PatternAlternative::PatternAlternative):

Canonical link: <a href="https://commits.webkit.org/290791@main">https://commits.webkit.org/290791@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b1eb3854a5c45aa05d443af1690323cf3f53036

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91077 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10621 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/113 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96085 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41848 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11030 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18941 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69999 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27521 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94078 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8400 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82531 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50325 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8171 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/95 "Found 1 new test failure: imported/w3c/web-platform-tests/cookies/schemeful-same-site/schemeful-websockets.sub.tentative.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40978 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/83899 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78493 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/112 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98064 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/89848 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18285 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79002 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18544 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78353 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78202 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22713 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/68 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14384 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18286 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23618 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/112416 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18012 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32642 "Found 9 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.mini-mode, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-slow-memory, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-slow-memory, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-collect-continuously, wasm.yaml/wasm/stress/repro_1289.js.wasm-eager, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch.js.default-wasm, wasm.yaml/wasm/stress/simple-inline-exception-inlinee-catch.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21473 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19791 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -1426,6 +1426,8 @@ class YarrPatternConstructor {
                 numBOLAnchoredAlts++;
         }
 
+        parenthesesDisjunction->m_alternatives[numParenAlternatives - 1]->m_isLastAlternative = true;
```
Is it possible that there is zero alternatives?
Also, we can use `last()`.

msaboff replied:
There should always be at least one alternative, so `last()` works.  I'll add an `ASSERT(numParenAlternatives)` above.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -1865,8 +1873,49 @@ class YarrPatternConstructor {
         Vector<std::unique_ptr<PatternAlternative>>& alternatives = m_pattern.m_body->m_alternatives;
-        for (size_t i = 0; i < alternatives.size(); ++i) {
-            Vector<PatternTerm>& terms = alternatives[i]->m_terms;
+        alternatives[alternatives.size() - 1]->m_isLastAlternative = true;
```
Ditto. We can also use `last()`.

msaboff replied:
Done.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -1865,8 +1873,49 @@ class YarrPatternConstructor {
+        }
+
+        for (size_t alternateNum = 0; alternateNum < alternatives.size(); ++alternateNum) {
+            Vector<PatternTerm>& terms = alternatives[alternateNum]->m_terms;
```
How about using
```
for (auto& alternative : alternatives) {
    auto& terms = alternative->m_terms;
...
```

msaboff replied:
Done.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -2826,6 +2898,9 @@ class YarrGenerator final : public YarrJITInfo {
         // Forwards generate the matching code.
         ASSERT(m_ops.size());
         size_t opIndex = 0;
+        Vector<MatchTargets> termMatchTargets;
```
Let's have some inline capacity like 8.

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -3737,19 +3875,28 @@ class YarrGenerator final : public YarrJITInfo {
+                    && m_backtrackingState.isEmpty())
+                    break;
+*/
+
```
We should remove this code?

msaboff replied:
Will do.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -5035,8 +5188,9 @@ class YarrGenerator final : public YarrJITInfo {
-                m_disassembler->dump(linkBuffer);
+            // Disassemble after all link tasks are complete.
+            m_jit.addLinkTask([=, this] (LinkBuffer& linkBuffer) {
+            m_disassembler->dump(linkBuffer);
```
Needs to have one level indentation.

msaboff replied:
Yes.

Constellation reviewed (approved):
r=me

msaboff committed:
[Yarr] Improve processing of an alternation of strings
https://bugs.webkit.org/show_bug.cgi?id=288102
rdar://145222010

Reviewed by Yusuke Suzuki.

Added the notion of a string list to a parsed RegExp that is in the form of
  /^(?:break|case|which|do|for)/ with an optional trailing $.
Such a RegExp will not backtrack and therefore we can streamline the code we emit for such a pattern.

This change involves recognizing beginning of string anchored alternations of strings while parsing and
then treating the generation of JIT code differently for these patterns.  This includes changing how
conditional branching works, specifically that instead of the "fall through on match" for each term,
to a "jump on match" for the whole alternation.

The current code generated for the "case" elternative is:
   8:Term PatternCharacter checked-offset:(3) 'c'
               <156> 0x11381430c:    add      w1, w1, #2
               <160> 0x113814310:    cmp      w1, w2
               <164> 0x113814314:    b.hi     0x113814444 -> <468>
  10:Term PatternCharacter checked-offset:(4) 'c'
               <168> 0x113814318:    sub      x17, x0, #4
               <172> 0x11381431c:    ldr      w17, [x17, x1]
               <176> 0x113814320:    movz     w16, #0x6163
               <180> 0x113814324:    movk     w16, #0x6573, lsl #16 -> 0x65736163
               <184> 0x113814328:    cmp      w17, w16
               <188> 0x11381432c:    b.ne     0x113814444 -> <468>
  11:Term PatternCharacter checked-offset:(4) 'a' already handled
  12:Term PatternCharacter checked-offset:(4) 's' already handled
  13:Term PatternCharacter checked-offset:(4) 'e' already handled
  14:NestedAlternativeNext minimum-size:(5),checked-offset:(5)
               <192> 0x113814330:    movz     x16, #0x4444
               <196> 0x113814334:    movk     x16, #0x1381, lsl #16
               <200> 0x113814338:    movk     x16, #0x8001, lsl #32
               <204> 0x11381433c:    movk     x16, #0xc973, lsl #48 -> 0x113814444 JIT PC
               <208> 0x113814340:    stur     x16, [sp, #8]
               <212> 0x113814344:    b        0x113814404 -> <404>
With some additional backtracking code:
   9:NestedAlternativeNext minimum-size:(4),checked-offset:(4)
               <468> 0x113814444:    sub      w1, w1, #2
               <472> 0x113814448:    b        0x113814348 -> <216>

With this change, the processing of "case" becomes:
   9:StringListAlternativeNext minimum-size:(4),checked-offset:(4)
               <132> 0x12a8285c4:    sub      w1, w1, #1
               <136> 0x12a8285c8:    cmp      w1, w2
               <140> 0x12a8285cc:    b.hi     0x12a8285e8 -> <168>
  10:Term PatternCharacter checked-offset:(4) 'c'
               <144> 0x12a8285d0:    sub      x17, x0, #4
               <148> 0x12a8285d4:    ldr      w17, [x17, x1]
               <152> 0x12a8285d8:    movz     w16, #0x6163
               <156> 0x12a8285dc:    movk     w16, #0x6573, lsl #16 -> 0x65736163
               <160> 0x12a8285e0:    cmp      w17, w16
               <164> 0x12a8285e4:    b.eq     0x12a82866c -> <300>
  11:Term PatternCharacter checked-offset:(4) 'a' already handled
  12:Term PatternCharacter checked-offset:(4) 's' already handled
  13:Term PatternCharacter checked-offset:(4) 'e' already handled
  14:StringListAlternativeNext minimum-size:(5),checked-offset:(5)
With no backtracking code.

We are able to eliminate one branch and the saving of the continuation PC for backtracking.
The code size to process these string list RegExp is reduces.  For the example RegExp above,
the prior version created 1940 bytes (485 instructions) of code while the code created with this
1392 bytes (345 instructions) of code, a nearly 30% reduction in code.

This change is a ~18% progression on the new regexp-keyword-parsing microbenchmark:

                                 Baseline               YarrStringList

regexp-keyword-parsing      136.7065+-0.9807     ^    116.0161+-1.1791        ^ definitely 1.1783x faster

<geometric>                 136.7065+-0.9807     ^    116.0161+-1.1791        ^ definitely 1.1783x faster

* JSTests/microbenchmarks/regexp-keyword-parsing.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
(let.re.break.case.catch.continue.debugger.default.else.finally.if):
(let.re1.break.case.catch.continue.debugger.default.else.finally.if):
* JSTests/stress/regexp-parsing-tokens.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::atomParenthesesEnd):
(JSC::Yarr::YarrPatternConstructor::checkForTerminalParentheses):
(JSC::Yarr::PatternAlternative::dump):
(JSC::Yarr::PatternTerm::dump):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::PatternTerm::PatternTerm):
(JSC::Yarr::PatternAlternative::PatternAlternative):

Canonical link: https://commits.webkit.org/290791@main
